src/Membrain/Memory.hs

```diff
@@ -39,6 +41,24 @@ newtype Memory (mem :: Nat) = Memory
     { unMemory :: Natural
     } deriving (Show, Read, Eq, Ord)
 
+instance Semigroup (Memory (mem :: Nat)) where
+    (<>) :: Memory mem -> Memory mem -> Memory mem
+    (<>) = coerce ((+) @Natural)
+    {-# INLINE (<>) #-}
+
+instance Monoid (Memory (mem :: Nat)) where
+    mempty :: Memory mem
+    mempty = Memory 0
+    {-# INLINE mempty #-}
+
+    mappend :: Memory mem -> Memory mem -> Memory mem
+    mappend = (<>)
+    {-# INLINE mappend #-}
+
+    mconcat :: [Memory mem] -> Memory mem
+    mconcat = foldl' (<>) mempty
```

Comment:
Nice usage of `foldl'`! This is indeed very efficient than the default implementation.